Lib/textwrap.py

```diff
@@ -199,6 +199,23 @@ def _fix_sentence_endings(self, chunks):
             else:
                 i += 1
 
+    def _find_width_index(self, text, width):
+        """_find_width_index(text : string, width: int)
+
+        Find at which index the text has the required width.
+        """
+        # In most cases text_len will just use the number of characters, so this heuristic prevents calculating width
+        # for each character
+        if self.text_len(text[:width]) == width:
+            # For character widths greater than one, width can be more than the number of characters
+            return min(width, len(text))
```

Comment:
I am not sure if this heuristic works. What would happen in the following case:

```python
lengths = {
    'A': 2,
    'B': 2,
}

def text_len(s):
    return sum(length.get(c, 1) for c in s)

wrap('AAAABBB', 3)
```

Reply:
If performance is an issue, a binary search might be an option.

Reply:
In that case the width of text[:width] (=['AAA']) is not equal to width (=3), so it will calculate it per character. However, if you'd used len as your text_len, you'd get that the width of ['AAA'] = 3, which is correct to it skips the calculation.

Binary search might work, but it'd require disallowing negative character lengths (which I agree don't make sense but the results of binary search might make the results more different than what you'd expect).